File: examples/r-api/query_lexicon.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)
query <- if (length(args) >= 1) args[[1]] else "agua"
limit <- if (length(args) >= 2) as.integer(args[[2]]) else 5L

if (is.na(limit) || limit < 1L || limit > 200L) {
  stop("El límite debe ser un entero entre 1 y 200.")
}

if (!requireNamespace("jsonlite", quietly = TRUE)) {
  stop("Falta la dependencia 'jsonlite'. Instálela con install.packages('jsonlite').")
}

base_url <- "https://raramuri.ceees.mx/api/lexicon"
url <- paste0(
  base_url,
  "?q=", utils::URLencode(query, reserved = TRUE),
  "&limit=", limit
)

payload <- tryCatch(
  jsonlite::fromJSON(url),
  error = function(e) stop("No fue posible consultar la API: ", conditionMessage(e))
)

entries <- payload$entries
if (is.null(entries) || NROW(entries) == 0L) {
  cat("No se encontraron resultados.\n")
  quit(status = 0)
}

fields <- intersect(c("recordId", "headword", "translationRaw"), names(entries))
print(entries[, fields, drop = FALSE], row.names = FALSE)

cat(
  "\nEstado: ", payload$publicationStatus,
  "; ", payload$validationStatus,
  "\n",
  sep = ""
)

File: examples/python-api/query_lexicon.py
#!/usr/bin/env python3
"""Consulta mínima reproducible de la API pública de Rarámuri Digital."""

from __future__ import annotations

import argparse
import json
import sys
import urllib.error
import urllib.parse
import urllib.request

API_URL = "https://raramuri.ceees.mx/api/lexicon"


def query_lexicon(query: str, limit: int) -> dict:
    params = urllib.parse.urlencode({"q": query, "limit": limit})
    request = urllib.request.Request(
        f"{API_URL}?{params}",
        headers={"Accept": "application/json", "User-Agent": "raramuri-digital-example/1.0"},
    )
    with urllib.request.urlopen(request, timeout=20) as response:
        if response.status != 200:
            raise RuntimeError(f"HTTP {response.status}")
        return json.load(response)


def main() -> int:
    parser = argparse.ArgumentParser(description="Consulta la API lexicográfica de Rarámuri Digital")
    parser.add_argument("q", nargs="?", default="agua", help="Texto de búsqueda (por defecto: agua)")
    parser.add_argument("--limit", type=int, default=5, help="Número de resultados, entre 1 y 200")
    args = parser.parse_args()

    if not 1 <= args.limit <= 200:
        parser.error("--limit debe estar entre 1 y 200")

    try:
        payload = query_lexicon(args.q, args.limit)
    except (urllib.error.URLError, TimeoutError, RuntimeError, json.JSONDecodeError) as exc:
        print(f"Error al consultar la API: {exc}", file=sys.stderr)
        return 1

    entries = payload.get("entries", [])
    if not entries:
        print("No se encontraron resultados.")
        return 0

    for entry in entries:
        print(
            f"{entry.get('recordId', '')}\t"
            f"{entry.get('headword', '')}\t"
            f"{entry.get('translationRaw', '')}"
        )

    print(
        f"\nEstado: {payload.get('publicationStatus', '')}; "
        f"{payload.get('validationStatus', '')}",
        file=sys.stderr,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
